x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -480,57 +525,95 @@ ClusterState createModelAssignment(ClusterState currentState, StartTrainedModelD
                     }
-                    rebalanceAssignments(currentState, modelToAdd, reason, listener);
-                    return currentState;
-                }
 
-                @Override
-                public void onFailure(Exception e) {
-                    listener.onFailure(e);
-                }
+                    @Override
+                    public void onFailure(Exception e) {
+                        listener.onFailure(e);
+                    }
 
-                @Override
-                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                    if (isUpdated) {
-                        if (isChanged) {
-                            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME)
-                                .execute(() -> systemAuditor.info(Messages.getMessage(Messages.INFERENCE_DEPLOYMENT_REBALANCED, reason)));
+                    @Override
+                    public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                        if (isUpdated) {
+                            if (isChanged) {
+                                threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME)
+                                    .execute(
+                                        () -> systemAuditor.info(Messages.getMessage(Messages.INFERENCE_DEPLOYMENT_REBALANCED, reason))
+                                    );
+                            }
+                            listener.onResponse(TrainedModelAssignmentMetadata.fromState(newState));
                         }
-                        listener.onResponse(TrainedModelAssignmentMetadata.fromState(newState));
                     }
-                }
+                });
             });
-        });
+        }, listener::onFailure);
+
+        MlPlatformArchitecturesUtil.getMlNodesArchitecturesSet(architecturesListener, client, threadPool);
+    }
+
+    ClusterState stopPlatformSpecificModelsInHeterogeneousClusters(
+        ClusterState updatedState,
+        Set<String> mlNodesArchitectures,
+        Optional<StartTrainedModelDeploymentAction.TaskParams> modelToAdd,
+        ClusterState clusterState
+    ) {
+        if (mlNodesArchitectures.size() > 1 && modelToAdd.isPresent()) {
+            String reasonToStop = format(
+                "ML nodes in this cluster have multiple platform architectures, "
+                    + "but can only have one for this model ([%s]); "
+                    + "detected architectures: %s",
+                modelToAdd.get().getModelId(),
+                mlNodesArchitectures
+            );
+            updatedState = callSetToStopping(reasonToStop, modelToAdd.get().getDeploymentId(), clusterState);
+        }
+        return updatedState;
+    }
+
+    ClusterState callSetToStopping(String reasonToStop, String deploymentId, ClusterState clusterState) {
+        return setToStopping(clusterState, deploymentId, reasonToStop);
```

Comment:
Just wanted to confirm that if multiple architectures are identified the model across all nodes is stopped?

Reply:
If there are any ML nodes in the cluster which have a different platform architecture from the one that is specified for the platform-specific-model, then the platform specific model must be stopped. So yes, if there are multiple architectures among ML nodes, any platform specific models will be stopped across all nodes.